actions/gha-workflow-validator/src/utils.ts

```diff
@@ -0,0 +1,165 @@
+      if (!dest.startsWith("+")) {
+        throw new Error("Invalid git hunk format");
+      }
+
+      const [ destLine, ] = dest.substring(1).split(',');
+      currentLineInFile = parseInt(destLine, 10);
+      continue;
+    } else if (line.startsWith('+')) {
+      const currentLine = line.substring(1);
+      const actionReference = extractActionReference(currentLine);
+      additions.push({ content: currentLine, lineNumber: currentLineInFile, actionReference });
+    } else if (line.startsWith('-')) {
+      // ignore deletions
+      continue;
+    }
+    currentLineInFile++;
+  }
+
+  return additions;
+}
+
+function extractActionReference(line: string): ActionReference | undefined {
+  // example line:
+  //       - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
+  const trimSubString = "uses:"
+  const usesIndex = line.indexOf(trimSubString);
+
+  if (usesIndex === -1) {
+    // Not an action reference
+    return;
+  }
+
+  // trim past the "uses:" substring to get "<owner>/<repo><optional path>@<ref> # <optional comment>""
+  const trimmedLine = line.substring(line.indexOf(trimSubString) + trimSubString.length).trim();
+
+  if (trimmedLine.startsWith("./")) {
+    // Local action reference - do not validate these yet (TODO)
+    return;
+  }
+
+  const [ actionIdentifier, ...comment ] = trimmedLine.split("#");
+  const [ identifier, gitRef ] = actionIdentifier.trim().split("@");
+  const [ owner, repo, ...path] = identifier.split("/");
+  const repoPath = ((path.length) > 0 ? "/" : "") + path.join("/");
+
+  return { owner, repo, repoPath, ref: gitRef, comment: comment.join().trim(), line, };
+}
+
+export function annotatePR(validationResults: ValidationResult[]) {
+  for (const fileResults of validationResults) {
+    for (const lineResults of fileResults.lineValidations) {
+      const message = lineResults.validationErrors.map(error => error.message).join("\n");
+      core.error(message, {
+        file: fileResults.filename,
+        startLine: lineResults.line.lineNumber,
+      });
+    }
+  }
+}
+
+type TableRow = Parameters<typeof core.summary.addTable>[0][0]
+type TableCell = TableRow[0]
+
+export async function setSummary(validationResults: ValidationResult[], fileUrlPrefix: string) {
+  const headerRows: TableRow[] = [
+    // Header Row
+    [
+      { data: "Filename", header: true },
+      { data: "Line Number", header: true },
+      { data: "Violations", header: true },
+    ],
+  ];
+
+  const errorRows = validationResults.reduce<TableRow[]>((acc, curr) => {
+    const filename = curr.filename;
+
+    const errorCellTuples: TableCell[][]  = curr.lineValidations.map(line => {
+      const lineNumberCell: TableCell = { data: htmlLink(`${line.line.lineNumber}`, `${fileUrlPrefix}/${filename}#L${line.line.lineNumber}`) }
+      const violationsCell: TableCell = { data: line.validationErrors.map(error => error.message).join(", ") }
+
+      return [ lineNumberCell, violationsCell ];
+    })
```

Comment:
```suggestion
    const errorCellTuples: TableCell[][]  = curr.lineValidations.map(({line}) => {
      const lineNumberCell: TableCell = { data: htmlLink(`${line.lineNumber}`, `${fileUrlPrefix}/${filename}#L${line.lineNumber}`) }
      const violationsCell: TableCell = { data: line.validationErrors.map(error => error.message).join(", ") }

      return [ lineNumberCell, violationsCell ];
    })
```

Apparently in go land this is called stuttering lmao

Reply:
This would break the code technically. I kinda want to figure out better typing/naming for these properties but I'm going to leave as is for now.

Can't destructure just to line because other properties are used.

```
line.line
line.validationErrors
```

Reply:
ah didnt notice